### PR TITLE
[Elastic Agent] Send Agent logs to elasticsearch

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Agent will now always log to `${path.data}/logs/elastic-agent-json.log` no matter the settings from `elastic-agent.yml` (those still apply, for the user specified choice). With monitoring always on Filebeat will always be started and will send the logs from `${path.data}/logs/elastic-agent-json.log` to elasticsearch.

The spawned filebeat and metricbeat now log to `filebeat-json.log` and `metricbeat-json.log` in JSON format and forward to elasticsearch.

Indexes:

* Elastic Agent - `logs-elastic.agent-default`
* Filebeat - `{logs,metrics}-elastic.agent.filebeat-default`
* Metricbeat - `{logs,metrics}-elastic.agent.metricbeat-default`

## Why is it important?

So all logs of the Elastic Agents on hosts are in elasticsearch so monitoring/troubleshooting of the Fleet is much easier.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates #16303

### CODE REVIEW

File: x-pack/elastic-agent/pkg/core/logger/logger.go [hunk 1]
```diff
@@ -99,10 +108,95 @@ func defaultConfig() (*common.Config, error) {
...
+		file.Interval(defaultCfg.Files.Interval),
+		file.RotateOnStartup(defaultCfg.Files.RotateOnStartup),
+		file.RedirectStderr(defaultCfg.Files.RedirectStderr),
+	)
+	if err != nil {
+		return nil, errors.New("failed to create internal file rotator")
+	}
+
+	encoder := zapcore.NewJSONEncoder(ecszap.ECSCompatibleEncoderConfig(logp.JSONEncoderConfig()))
+	return ecszap.WrapCore(zapcore.NewCore(encoder, rotator, logp.DebugLevel.ZapLevel())), nil
```
Comment (urso) [inline x-pack/elastic-agent/pkg/core/logger/logger.go:140]: I wonder if we could have reduced some of the plumbing by exporting `logp.createLogOutput only` and split configure into `configure log output` and `configure logging with output`. The `zapcore.Core` is basically the output here. The core that is produced here coulbe be build via `CreateLogOutput(agentConfig)`.

Allowing us to pass a separate output (`zapcore.Core`) to `logp.Configure` might allow us in the future to change how logging will be configured in other products, but still work with existing code in Beats that requires logp to function as is (e.g. for agent or collector).

**Comment by blakerouse** [comment]
@urso I am trying to make as little change as possible, don't want to cause a big refactor on FF day. I have added a `ConfigureWithOutputs` see if that is good enough for now.

File: libbeat/logp/core.go [hunk 2]
```diff
@@ -62,6 +64,12 @@ type coreLogger struct {
 
 // Configure configures the logp package.
 func Configure(cfg Config) error {
+	return ConfigureWithOutputs(cfg)
+}
+
+// Configure configures the logp package also sets up the following outputs to also be used
+// by the logger.
+func ConfigureWithOutputs(cfg Config, outputs ...zapcore.Core) error {
```
Comment (urso) [inline libbeat/logp/core.go:72]: Let's add a note here that agent expects `ConfigureWithOutputs` to still create the output. When cleaning up/refactoring this function in the method I would move the output init to the `Configure` method, which can break logging in the agent from working correctly.

e.g.

```
func Configure(cfg Config) error {
  logOutput, err := CreateLogOutput(cfg.X)
  if err != nil { ... }
  return ConfigureWithOutputs(cfg, logOutput)
}
```

**Comment by blakerouse** [comment]
@urso Added comment.

File: libbeat/logp/core.go [hunk 3]
```diff
@@ -67,8 +67,9 @@ func Configure(cfg Config) error {
 	return ConfigureWithOutputs(cfg)
 }
 
-// Configure configures the logp package also sets up the following outputs to also be used
-// by the logger.
+// Configure configures the logger from passed in config and also adds the passed outputs to the logger.
+//
+// When passing in outputs the logger will log to each output.
```
Comment (urso) [inline libbeat/logp/core.go:72]: ConfigureWithOutputs is currently used by agent only. We have a dependency between this function and what the agent expects. Let's add a line saying:

```
// XXX: ConfigureWithOutputs is used by elastic-agent only (See file: ....). The agent requires that the output specified in the config object is confitured and merged with the logging outputs given. 
```

Comment (blakerouse) [reply libbeat/logp/core.go]: Updated comment to match suggestion.

**Pull request merged**